components/server/src/test/service-testing-container-module.ts

```diff
@@ -4,44 +4,70 @@
+import { StorageClient } from "../storage/storage-client";
+import { IDEServiceClientMock } from "./mocks/ide-service-mock";
+import { StorageClientMock } from "./mocks/storage-client-mock";
+import { UsageServiceClientMock } from "./mocks/usage-service-client-mock";
+import { WorkspaceManagerClientProviderMock } from "./mocks/workspace-manager-client-mock";
+import { testContainer } from "@gitpod/gitpod-db/lib";
+import { productionContainerModule } from "../container-module";
+
+/**
+ * Expects a fully configured production container and
+ *  - replaces all services to external APIs with mocks
+ *  - replaces the config with a mock config
+ *  - replaces the analytics writer with a null analytics writer
+ */
+const mockApplyingContainerModule = new ContainerModule((bind, unbound, isbound, rebind) => {
+    rebind(UsageServiceDefinition.name).to(UsageServiceClientMock).inSingletonScope();
+    rebind(StorageClient).to(StorageClientMock).inSingletonScope();
+    rebind(WorkspaceManagerClientProvider).toConstantValue(new WorkspaceManagerClientProviderMock());
+    rebind(IDEServiceDefinition.name).to(IDEServiceClientMock).inSingletonScope();
 
-export const serviceTestingContainerModule = new ContainerModule((bind) => {
-    bind(OrganizationService).toSelf().inSingletonScope();
-    bind(ProjectsService).toSelf().inSingletonScope();
-    bind(Config).toConstantValue({});
-    bind(AuthProviderService).toSelf().inSingletonScope();
-    bind(IAnalyticsWriter).toConstantValue(NullAnalyticsWriter);
-    // hostcontext
-    bind(HostContainerMapping).toSelf().inSingletonScope();
-    bind(HostContextProviderFactory)
+    rebind<Partial<Config>>(Config).toConstantValue({
+        blockNewUsers: {
+            enabled: false,
+            passlist: [],
+        },
+        redis: {
+            address: "localhost:6379",
+        },
+    });
+    rebind(IAnalyticsWriter).toConstantValue(NullAnalyticsWriter);
+    rebind(HostContextProviderFactory)
         .toDynamicValue(({ container }) => ({
             createHostContext: (config: AuthProviderParams) =>
                 HostContextProviderImpl.createHostContext(container, config),
         }))
         .inSingletonScope();
-    bind(HostContextProvider).to(HostContextProviderImpl).inSingletonScope();
 
-    bind(SpiceDBClient)
+    rebind(SpiceDBClient)
         .toDynamicValue(() => {
             const token = v4();
             return v1.NewClient(token, "localhost:50051", v1.ClientSecurity.INSECURE_PLAINTEXT_CREDENTIALS).promises;
         })
         .inSingletonScope();
-    bind(SpiceDBAuthorizer).toSelf().inSingletonScope();
-    bind(Authorizer).toSelf().inSingletonScope();
 });
+
+/**
+ *
+ * @returns a container that is configured for testing and assumes a running DB, spiceDB and redis
+ */
+export function createTestContainer() {
+    const container = testContainer.createChild();
+    container.load(productionContainerModule);
+    container.load(mockApplyingContainerModule);
```

Comment:
We now use the production container module but cut off all dependencies to external services with the mockApplyingContainer Module. So we essentially get a full server component as in production but with mocks on the outer ends. This keeps the testing setup and production setup aligned and also tests the production container module.